### ISSUE TITLE
Steam web requests sometimes fail

### ISSUE THREAD

**Issue by azalty** [azalty/sm-no-dupe-account#12]
**Describe the bug**
Steam web requests sometimes fail:
[no_dupe_account.smx] Steam Age request failed!
[no_dupe_account.smx] Steam Bans request failed!
[no_dupe_account.smx] Steam Level request failed!
[no_dupe_account.smx] Playtime request failed!

They often happen at the same time. Couldn't find the reason why.

~~It seems that Steam Age and Steam Level requests sometimes fail to grab the user's level/account age, and display '0' (string -> int conversion failed). Need more info.~~

~~A pretty consequent amount of steam web requests seem to fail as well, for some reason.~~

**Comment by azalty** [comment]
I am not able to replicate this problem myself, I'm creating a debug system so I can try to understand the problem

**Comment by rcon420** [comment]
I noticed this once. Few guys got kicked all the time and next map they were able to connect again. In fact tho they didn't pass all the checks but passed some others. I haven't noticed this problem because this requires seeing the server console.

**Comment by azalty** [comment]
> I noticed this once. Few guys got kicked all the time and next map they were able to connect again. In fact tho they didn't pass all the checks but passed some others. I haven't noticed this problem because this requires seeing the server console.

Could you please send me those logs, or those saying that web requests failed?

As of now I still don't know what causes these fails, but they only seem to happen with steam requests and not VPN requests.

I'll try to have a look at my own server since I added debug for that

Is that linked to #15?

**Comment by rcon420** [comment]
> > I noticed this once. Few guys got kicked all the time and next map they were able to connect again. In fact tho they didn't pass all the checks but passed some others. I haven't noticed this problem because this requires seeing the server console.
> 
> Could you please send me those logs, or those saying that web requests failed?
> 
> As of now I still don't know what causes these fails, but they only seem to happen with steam requests and not VPN requests.
> 
> I'll try to have a look at my own server since I added debug for that
> 
> Is that linked to #15?

I had this problem before the Database thingy. Sorry, I don't have that old logs anymore.

**Comment by azalty** [comment]
If they passed at least one check then the plugin behaves normally (assuming the web request didn't fail and let them enter). In case they didn't pass at least one requirement, and still got accepted, that's a problem. (See "What's a check" on the wiki)

I'm starting to think it's due to the fact that multiple web requests to steamapi are sent the same time, and for some reason when getting multiple requests at the same time, steam could deny the request (for example on map change -> every player is checked again at the same time) or steamworks could fuck up and not get the webpage correctly

Alternatively, steam api could be overloaded or down (not sure about that one)

**Comment by azalty** [comment]
I fucked up, Steam Level 0 exists, I'm modifying my own test version to see if these problems still occur

**Comment by azalty** [comment]
I think the conversion error is linked to requests error (or maybe not)

I'm changing this issue's topic. You should create a new issue if you think these problems aren't related.

About requests failing: I got 60 request fails in 7 days on my servers, which is kind of a lot. Some appear alone, others in groups for a few mins

**Comment by azalty** [comment]
These kind of problems still seem to occur.

Web requests to the Steam API that fail during Steam weekly's maintenances is **normal**, but fails for the Steam API still occur outside of maintenances.

If web requests fail one minute after another, it could be a rate-limit Steam has implemented. I still have no idea why those fail.

I'll push another commit to the beta version soon to add a bit more logging (retrieve the HTTP Status Code)

**Comment by crashzk** [comment]
Apparently with the latest version this problem stopped occurring, today it occurred but it was during the steam maintenance window, before this maintenance there are no more errors.

**Comment by azalty** [comment]
> Apparently with the latest version this problem stopped occurring, today it occurred but it was during the steam maintenance window, before this maintenance there are no more errors.

Still occurs for me.
I'll write an experimental patch to retry the request after a few seconds, might fix the problem.

**Comment by azalty** [comment]
Added HTTP Request retries
A lot of the code for the web requests have been changed. I didn't get any error or bug while testing, but watch out.

By default it will do 4 tries (1 original + 3 retries), each being delayed by 5s.
Let's hope this fixes the problem

**Comment by azalty** [comment]
The retry system seems working so far!
Only got these two fails, and they worked on the 2nd try:

Note: It was status code 503 which means `ServiceUnavailable`. I usually get Status Code 0 `Invalid/Unspecified`, but it's nice to see that this fail got handled. I'll give more infos about Status Code 0 fails as soon as I get one.

```
L 11/12/2021 - 22:59:24: [no_dupe_account.smx] QLF (STEAM_1:1:97868429) database status: Checked and should be updated
Values: csgo_level=20, csgo_coin=874, prime=1, csgo_playtime=-1 (minutes), steam_level=-1, steam_age=-1 (minutes), last_check=1636307146 (timestamp)
L 11/12/2021 - 22:59:24: [no_dupe_account.smx] Approved QLF<145><STEAM_1:1:97868429><> (Is Prime)
L 11/12/2021 - 22:59:26: [no_dupe_account.smx] Steam Age request failed! - Status Code: 503 - Try 1/4
L 11/12/2021 - 22:59:26: [no_dupe_account.smx] Steam Level request failed! - Status Code: 503 - Try 1/4
```

**Comment by azalty** [comment]
Seems working for Status Code 0 too (no 2nd try fail)
```
L 11/19/2021 - 22:07:15: [no_dupe_account.smx] Playtime request failed! - Status Code: 0 - Try 1/4
L 11/19/2021 - 22:07:15: [no_dupe_account.smx] Steam Age request failed! - Status Code: 0 - Try 1/4
L 11/20/2021 - 18:09:10: [no_dupe_account.smx] Playtime request failed! - Status Code: 0 - Try 1/4
L 11/21/2021 - 10:05:04: [no_dupe_account.smx] Steam Bans request failed! - Status Code: 0 - Try 1/4
```

Seems like the Steam API/services were not available. It failed even after multiple retries. It's safe to assume that this problem if from Steam, and not NDA.
```
L 11/20/2021 - 02:03:39: [no_dupe_account.smx] Steam Bans request failed! - Status Code: 503 - Try 1/4
L 11/20/2021 - 02:03:39: [no_dupe_account.smx] Steam Bans request failed! - Status Code: 503 - Try 1/4
L 11/20/2021 - 02:03:45: [no_dupe_account.smx] Steam Bans request failed! - Status Code: 503 - Try 2/4
L 11/20/2021 - 02:03:45: [no_dupe_account.smx] Steam Bans request failed! - Status Code: 503 - Try 2/4
L 11/20/2021 - 02:03:51: [no_dupe_account.smx] Steam Bans request failed! - Status Code: 503 - Try 3/4
L 11/20/2021 - 02:03:52: [no_dupe_account.smx] Steam Bans request failed! - Status Code: 503 - Try 3/4
L 11/20/2021 - 02:03:57: [no_dupe_account.smx] Steam Bans request failed! - Status Code: 503 - Try 4/4
L 11/20/2021 - 02:03:58: [no_dupe_account.smx] Steam Bans request failed! - Status Code: 503 - Try 4/4
```

This is now considered fixed in beta. If any problem still happens, please report it :)

**Comment by crashzk** [comment]
Perfect, I believe this issues can be closed, I will update to the latest version of the Beta, any problem I will contact.

**Comment by azalty** [comment]
Fix added in the 1.5.0 stable release.

If there's any problem, comment & re-open this issue :)
